api/v1beta1/machinedrainrules_types.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// PodDrainLabel is the label that can be set on Pods in workload clusters to ensure a Pod is not drained.
+	// The only valid value is "skip".
+	// This label takes precedence over MachineDrainRules defined in the management cluster.
+	PodDrainLabel = "cluster.x-k8s.io/drain"
+)
+
+// MachineDrainRuleDrainBehavior defines the drain behavior. Can be either "Drain" or "Skip".
+// +kubebuilder:validation:Enum=Drain;Skip
+type MachineDrainRuleDrainBehavior string
+
+const (
+	// MachineDrainRuleDrainBehaviorDrain means a Pod should be drained.
+	MachineDrainRuleDrainBehaviorDrain MachineDrainRuleDrainBehavior = "Drain"
+
+	// MachineDrainRuleDrainBehaviorSkip means the drain for a Pod should be skipped.
+	MachineDrainRuleDrainBehaviorSkip MachineDrainRuleDrainBehavior = "Skip"
+)
+
+// MachineDrainRuleSpec defines the spec of a MachineDrainRule.
+type MachineDrainRuleSpec struct {
+	// drain configures if and how Pods are drained.
+	// +required
+	Drain MachineDrainRuleDrainConfig `json:"drain"`
+
+	// machines defines to which Machines this MachineDrainRule should be applied.
+	//
+	// If machines is empty, the MachineDrainRule applies to all Machines in the Namespace.
```

Comment:
Is this contradictory to this field being required?

Reply:
Hm. I was thinking it's possible to set this field to `[]`. But I guess we now broke this with MinItems. I'll play around a bit with it

Reply:
Continued in https://github.com/kubernetes-sigs/cluster-api/pull/11353#discussion_r1824285984

Reply:
This comment is gone now